Microsoft.NET.Build.Containers/Registry.cs

```diff
@@ -82,39 +98,112 @@ public Registry(Uri baseUri)
+        var config = manifest.config;
+        string configSha = config.digest;
+        
+        var blobResponse = await GetBlob(repositoryName, configSha);
 
-        response.EnsureSuccessStatusCode();
+        JsonNode? configDoc = JsonNode.Parse(await blobResponse.Content.ReadAsStringAsync());
+        Debug.Assert(configDoc is not null);
 
-        var s = await response.Content.ReadAsStringAsync();
+        return new Image(manifest, configDoc, repositoryName, this);
+    }
 
-        var manifest = JsonNode.Parse(s);
+    async Task<Image?> TryPickBestImageFromManifestList(string repositoryName, string reference, ManifestListV2 manifestList, string runtimeIdentifier, string runtimeIdentifierGraphPath) {
+        var runtimeGraph = GetRuntimeGraphForDotNet(runtimeIdentifierGraphPath);
+        var (ridDict, graphForManifestList) = ConstructRuntimeGraphForManifestList(manifestList, runtimeGraph);
+        var bestManifestRid = CheckIfRidExistsInGraph(graphForManifestList, runtimeIdentifier);
+        if (bestManifestRid is null) {
+            throw new ArgumentException($"The runtimeIdentifier '{runtimeIdentifier}' is not supported. The supported RuntimeIdentifiers for the base image {repositoryName}:{reference} are {String.Join(",", graphForManifestList.Runtimes.Keys)}");
+        }
+        var matchingManifest = ridDict[bestManifestRid];
+        var manifestResponse = await GetManifest(repositoryName, matchingManifest.digest);
+        return await TryReadSingleImage(repositoryName, await manifestResponse.Content.ReadFromJsonAsync<ManifestV2>());
+    }
 
-        if (manifest is null) throw new NotImplementedException("Got a manifest but it was null");
+    private async Task<HttpResponseMessage> GetManifest(string repositoryName, string reference)
+    {
+        var client = GetClient();
+        var response = await client.GetAsync(new Uri(BaseUri, $"/v2/{repositoryName}/manifests/{reference}"));
+        response.EnsureSuccessStatusCode();
+        return response;
+    }
 
-        if ((string?)manifest["mediaType"] != DockerManifestV2)
-        {
-            throw new NotImplementedException($"Do not understand the mediaType {manifest["mediaType"]}");
-        }
+    async Task<HttpResponseMessage> GetBlob(string repositoryName, string digest)
+    {
+        var client = GetClient();
+        var response = await client.GetAsync(new Uri(BaseUri, $"/v2/{repositoryName}/blobs/{digest}"));
+        response.EnsureSuccessStatusCode();
+        return response;
+    }
 
-        JsonNode? config = manifest["config"];
-        Debug.Assert(config is not null);
-        Debug.Assert(((string?)config["mediaType"]) == DockerContainerV1);
+    private string? CheckIfRidExistsInGraph(RuntimeGraph graphForManifestList, string userRid) => graphForManifestList.Runtimes.FirstOrDefault(kvp => graphForManifestList.AreCompatible(kvp.Key, userRid)).Key;
```

Comment:
Is there guaranteed to be only one compatible RID? If not, enumerating a dictionary could result in changes here right?